### PR TITLE
feat(meta): SEAL scaffolding aligned with architecture (Fixes #289)

### LINKED ISSUE #289: Implement SEAL Meta-Learning Algorithm (Core Integration)
Title: Implement SEAL Meta-Learning Algorithm (Core Integration)

Background
- SEAL (MIT-affiliated, official repo: https://github.com/Continual-Intelligence/SEAL) proposes a self-improving meta-learning approach enabling rapid adaptation from experience.
- Investigation issue (#286) tracks literature/code review and integration planning.

Problem
- AiDotNet lacks a production-grade meta-learning algorithm for few-shot adaptation across tasks.

User Story
- As a researcher/practitioner, I want a first-class SEAL implementation with episodic data pipelines, so I can train and evaluate rapid adaptation on standard few-shot benchmarks and integrate with existing models.

Scope
- Implement SEAL core algorithm and training loop abstractions.
- Provide episodic datasets/dataloaders, task samplers, and evaluation harness.
- Benchmark vs baseline meta-learners (MAML/iMAML/Reptile/Meta-SGD/ALFA).

Acceptance Criteria
- Algorithms
  - SEAL core algorithm implemented with clear separation of inner/outer updates and adaptation steps.
  - Configurable hyperparameters (inner steps, learning rates, adaptation schedule, temperature/regularization terms).
- Data & Tasks
  - Episodic dataset/dataloader interfaces (`N-way K-shot` tasks) with reproducible samplers.
  - Reference tasks: Omniglot and MiniImageNet few-shot classification; simple regression meta-task (optional).
- Training & Eval
  - `MetaTrainer` with resume/checkpointing, deterministic seeding, and logging.
  - Evaluation protocol: meta-train/meta-val/meta-test splits; report accuracy across episodes with confidence intervals.
  - Baseline comparisons: MAML/Reptile (and at least one of iMAML/Meta-SGD/ALFA) trained and evaluated with same splits.
- APIs & Integration
  - Namespaces: `AiDotNet.MetaLearning.{Algorithms,Tasks,Samplers,Training}`.
  - Pluggable base learners (e.g., small ConvNet/ViT) with adapter interface.
- Tests
  - Unit tests: inner/outer loop correctness on synthetic tasks; sampler determinism.
  - E2E smoke test: 5-way 1-shot Omniglot run reaching non-trivial accuracy above random.
- Docs & Examples
  - Guide: “Train SEAL on MiniImageNet/Omniglot.”
  - Example scripts and configs.
- License & Provenance
  - Verify compatibility with SEAL repo license; attribute appropriately in docs.

Milestones
- M1 – Design Sign-off (2–3 days)
  - Finalize integration plan from #286, define public APIs, and list datasets.
- M2 – Episodic Data Abstractions (1 week)
  - Implement datasets/dataloaders and samplers; add tests.
- M3 – SEAL Core Implementation (1–2 weeks)
  - Implement algorithm, config, and training loop; add unit tests.
- M4 – Baselines & Evaluation (1 week)
  - Implement at least two baselines (MAML/Reptile), run benchmarks, and compare metrics.
- M5 – Documentation & Examples (3–5 days)
  - Add user guide, examples, and API docs.
- M6 – Hardening & CI (3–5 days)
  - E2E smoke tests; regression checks; polish.

Dependencies
- #286 Meta-Learning Suite and SEAL Investigation.
- Datasets framework (see 022) – can bootstrap with local loaders if needed.

Risks & Mitigations
- Training instability: start with small models and lower shot/way; checkpoint frequently; seed deterministically.
- Dataset licensing: ensure permissible distribution/instructions.

Deliverables
- Working SEAL implementation and training scripts.
- Benchmarks with report (accuracy/CIs) vs baselines.
- Docs, examples, and tests merged.

### PR DESCRIPTION
**Pull request by ooples**
Scaffolds SEAL meta-learning integration aligned with project architecture (generic T, Tensor<T>, Vector<T>, INumericOperations<T>), to prioritize #289.

- Adds IMetaLearner<T,TInput,TOutput> and ISEALModel<T,TInput,TOutput>
- Adds initial SEALModel<T> implementing IFullModel<T, Tensor<T>, Tensor<T>>; methods stubbed for follow-up commits
- Adds minimal SEALModelTests to validate construction and prediction shape

Next steps in this PR
- Implement SEAL episodic training loop (support/query) with tests
- Integrate with PredictionModelBuilder<T,TInput,TOutput>.ConfigureModel(...)
- Add metrics and evaluation hooks consistent with existing evaluators

Fixes #289 (initial scaffolding)

### CODE REVIEW

**Comment by ooples** [comment]
## Closing Summary

Thank you for this SEAL scaffolding PR. After comprehensive architectural analysis comparing this PR, PR #307, and master against industry standards, we have determined that while this PR contains valuable integration patterns, some architectural decisions need refinement before adoption.

## Architectural Analysis

### Components to ADOPT

1. **IMetaLearningStrategy<T, TInput, TOutput>**: ✅ Excellent abstraction
   - Clean integration hook into PredictionModelBuilder workflow
   - Prepare method provides clear interception point for OptimizationInputData
   - TryPredict method enables inference-time adaptation
   - This will be the primary integration point for all meta-learning algorithms

2. **PredictionModelBuilder.ConfigureMetaLearning()**: ✅ Good pattern
   - Proper builder method for strategy injection
   - Maintains fluent API consistency

### Components to REJECT

1. **IMetaLearner<T, TInput, TOutput> inheriting IFullModel**: ❌ Architectural mismatch
   - Meta-learner typically *trains* a base IFullModel, not *is* the model
   - Predict/Train methods of IFullModel don't align with meta-learning inner/outer loops
   - Conflation of roles leads to awkward API

2. **SEALModel<T> and ISEALModel<T, TInput, TOutput>**: ❌ Inherits mismatch
   - Should use composition instead of inheritance
   - SEALAlgorithm should operate *on* an IFullModel, not *be* an IFullModel

## Recommended Approach (from Issue #442)

Instead of IMetaLearner as IFullModel, we will implement:

```csharp
// Correct architecture
public class SEALMetaLearningStrategy<T, Tensor<T>, Tensor<T>> 
    : IMetaLearningStrategy<T, Tensor<T>, Tensor<T>>
{
    private readonly SEALConfig<T> _config;
    private readonly IFullModel<T, Tensor<T>, Tensor<T>> _baseModel;  // Composition
    private readonly SEALAlgorithm<T> _algorithm;  // IMetaLearner<T> from PR #307

    public OptimizationInputData Prepare(OptimizationInputData data)
    {
        // Use ITaskSampler<T> to generate IEpisode<T> batches
        // Call _algorithm.MetaTrainStep() to update _baseModel
        // Return updated OptimizationInputData
    }

    public TOutput TryPredict(TInput input, IEpisode<T> supportSet)
    {
        // Adapt _baseModel for new task
        // Return prediction
    }
}
```

## Why Closing

1. **Architectural Refinement Needed**: IMetaLearner as IFullModel pattern needs correction
2. **Better Integration Path**: Use IMetaLearningStrategy (which you provided!) with composition
3. **Coordinated with PR #307**: SEAL implementation needs data abstractions from PR #307

## Next Steps

Your IMetaLearningStrategy abstraction will be adopted as the primary integration mechanism. The SEAL algorithm will be implemented following the corrected architecture outlined in **Issue #442: Adopt Industry-Standard Meta-Learning Data Abstractions from PR #307 and PR #299**.

Thank you for the excellent IMetaLearningStrategy pattern - this is exactly the right integration point!

**Related**: Issue #442, PR #307
